### ISSUE TITLE
ECS Volumes Are Incorrectly Defined As Arguments Instead of Blocks (No Equal Sign Allowed)

### ISSUE THREAD

**Issue by grahamschuckman** [techservicesillinois/terraform-aws-ecs-service#53]
Volumes are defined with equal signs in your code which is not permitted by Terraform. Volumes for ECS have to be defined as blocks just using the {}. See documentation for reference: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_task_definition

  volume {
    name      = "service-storage"
    host_path = "/ecs/service-storage"
  }
  
The same is true for any configuration blocks inside your volume block, such as efs_volume_configuration:

resource "aws_ecs_task_definition" "service" {
  family                = "service"
  container_definitions = file("task-definitions/service.json")

  volume {
    name = "service-storage"

    efs_volume_configuration {
      file_system_id          = aws_efs_file_system.fs.id
      root_directory          = "/opt/data"
      transit_encryption      = "ENABLED"
      transit_encryption_port = 2999
      authorization_config {
        access_point_id = aws_efs_access_point.test.id
        iam             = "ENABLED"
      }
    }
  }
}

**Comment by JonRoma** [comment]
@grahamschuckman, sorry that the issue you reported languished in our queue for 10 months!

If you are still having this problem, I would be interested to know what version you are using. I'm looking at v3.2.1 (the latest production release) and do not see an issue. Moreover, we have services using the volume block in production code, so I'm certain that it works.

```
  volume = [
    {
      name                        = "efs-idp-configs"
      host_path                   = null
      docker_volume_configuration = null

      efs_volume_configuration = {
        file_system_id = "fs-****************"
        root_directory = "/shibboleth-idp"
      }
    }
  ]
```

You might be interested to know that a major refactor just went to beta this morning; it eliminates a lot of legacy code required to circumvent limitations in ancient versions of Terraform.

**Comment by JonRoma** [comment]
@grahamschuckman, I've examined this again.

> Volumes are defined with equal signs in your code which is not permitted by Terraform. Volumes for ECS have to be defined as blocks just using the {}. 

Let's take the `var.volume` that you mention. This input variable is declared as a list of objects in the module's `variables.tf`. In `task.tf`, we iterate over this list using a `for_each` inside a `dynamic` block to create one `volume` block for each volume definition in the variable.

This code was merged during June 2020 in commit dec26a46, which is tagged as version v3.1.0. As mentioned, the `volume` block works as designed in numerous infrastructure definitions that are used in production here.

I'm closing this issue, but you are welcome to reopen it (or create a new issue), providing a concrete example of what you believe doesn't work. Thanks for your feedback.
